presto-main/src/main/java/com/facebook/presto/operator/LongArrayBlockEncodingBuffers.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airlift.slice.SliceOutput;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.operator.UncheckedByteArrays.setLong;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.Math.max;
+import static sun.misc.Unsafe.ARRAY_LONG_INDEX_SCALE;
+
+public class LongArrayBlockEncodingBuffers
+        extends BlockEncodingBuffers
+{
+    @VisibleForTesting
+    public static final int POSITION_SIZE = Long.BYTES + Byte.BYTES;
+
+    private static final String NAME = "LONG_ARRAY";
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(LongArrayBlockEncodingBuffers.class).instanceSize();
+
+    private byte[] valuesBuffer;
+    private int valuesBufferIndex;
+
+    public LongArrayBlockEncodingBuffers(int initialPositionCount)
+    {
+        super(initialPositionCount);
+        prepareBuffers();
+    }
+
+    @Override
+    protected void prepareBuffers()
+    {
+        valuesBuffer = new byte[initialPositionCount * ARRAY_LONG_INDEX_SCALE];
+        valuesBufferIndex = 0;
+    }
+
+    @Override
+    public void resetBuffers()
+    {
+        bufferedPositionCount = 0;
+        valuesBufferIndex = 0;
+        resetNullsBuffer();
+    }
+
+    @Override
+    public void accumulateRowSizes(int[] rowSizes)
+    {
+        throw new UnsupportedOperationException("accumulateRowSizes is not supported for fixed width types");
+    }
+
+    @Override
+    protected void accumulateRowSizes(int[] positionOffsets, int positionCount, int[] rowSizes)
+    {
+        for (int i = 0; i < positionCount; i++) {
+            rowSizes[i] += (positionOffsets[i + 1] - positionOffsets[i]) * POSITION_SIZE;
+        }
+    }
+
+    @Override
+    public void copyValues()
+    {
+        if (batchSize == 0) {
```

Comment:
When would this be the case?

Reply:
The positions in nested level could be null. In this case the batchSize is 0.

Reply:
Got it. Thanks for clarifying.